src/relax/backend/vm/rewrite_cuda_graph.cc

```diff
@@ -0,0 +1,410 @@
+    extractor.VisitExpr(func->body);
+    return extractor.block_to_static_regions_;
+  }
+  StaticRegionExtractor(const IRModule& mod, const std::unordered_set<const VarNode*>& outputs)
+      : mod_(mod), outputs_(outputs) {}
+
+  struct DependencyGraph {
+    void AddBinding(const VarBindingNode* binding, bool is_alloc_tensor) {
+      const auto* var = binding->var.get();
+      if (is_alloc_tensor) {
+        alloc_tensors.push_back(var);
+      } else {
+        nodes.push_back(var);
+      }
+      bindings.emplace(var, binding);
+    }
+
+    void AddEdge(const VarNode* src, const VarNode* dst) {
+      dep_src2dst[src].push_back(dst);
+      dep_dst2src[dst].push_back(src);
+    }
+
+    std::unordered_map<const VarNode*, std::vector<const VarNode*>> dep_src2dst;
+    std::unordered_map<const VarNode*, std::vector<const VarNode*>> dep_dst2src;
+    std::vector<const VarNode*> alloc_tensors;
+    std::vector<const VarNode*> nodes;
+    std::unordered_map<const VarNode*, const VarBindingNode*> bindings;
+  };
+
+  struct ScopeInfo {
+    DependencyGraph graph;
+    std::vector<StaticRegion> static_regions;
+  };
+
+  void VisitBindingBlock_(const BindingBlockNode* block) final {
+    ScopeInfo scope;
+    std::swap(scope, scope_);
+    for (const auto& binding : block->bindings) {
+      VisitBinding(binding);
+    }
+    SummarizeStaticRegion();
+    std::swap(scope, scope_);
+    if (scope.static_regions.size()) {
+      block_to_static_regions_.emplace(block, std::move(scope.static_regions));
+    }
+  }
+
+  void VisitBinding_(const MatchCastNode* binding) final {
+    ExprVisitor::VisitBinding_(binding);
+    SummarizeStaticRegion();
+  }
+
+  bool IsStaticAlloc(const CallNode* alloc_tensor_call) {
+    auto shape = Downcast<ShapeExpr>(alloc_tensor_call->args[0]);
+    return std::all_of(shape->values.begin(), shape->values.end(),
+                       [](const PrimExpr& expr) { return expr.as<IntImmNode>() != nullptr; });
+  }
+
+  void VisitBinding_(const VarBindingNode* binding) final {
+    ExprVisitor::VisitBinding_(binding);
+
+    const Expr& value = binding->value;
+    static const auto& alloc_tensor_op = Op::Get("relax.builtin.alloc_tensor");
+    if (value->IsInstance<VarNode>()) {
+      // var -> var re-binding is not needed as we expect the binding is canonicalized
+      return;
+    }
+    if (value->IsInstance<ConstantNode>()) {
+      scope_.graph.AddBinding(binding, false);
+      return;
+    }
+    if (const auto* call = value.as<CallNode>()) {
+      if (call->op == alloc_tensor_op && IsStaticAlloc(call)) {
+        scope_.graph.AddBinding(binding, true);
+        return;
+      } else if (const auto* gv = call->op.as<GlobalVarNode>()) {
+        auto func = mod_->Lookup(GetRef<GlobalVar>(gv));
+        if (!func->IsInstance<tir::PrimFuncNode>()) {
+          return;
+        }
+        bool is_all_args_static =
+            std::all_of(call->args.begin(), call->args.end(), [&](const Expr& arg) -> bool {
```

Comment:
Can it be a Tuple of static arguments, i.e. Tuple of ints?

Reply:
good points